posts/2020-05-15-five-years-of-rust.md

```diff
@@ -0,0 +1,497 @@
+src/lib.rs:6     generic(w);
+                 ^~~~~~~
+src/lib.rs:6:5: 6:12 note: `std::io::Write` does not have a constant size known at compile-time
+src/lib.rs:6     generic(w);
+                 ^~~~~~~
+error: aborting due to previous error
+Could not compile `error-messages`.
+
+To learn more, run the command again with --verbose.
+```
+
+</details>
+
+![A terminal screenshot of the 1.2.0 error message.][trait-error-1.2.0]
+
+<details>
+ <summary>1.43.0 Error Message</summary>
+
+```
+   Compiling error-messages v0.1.0 (/Users/ep/src/rust/error-messages)
+error[E0277]: the size for values of type `dyn std::io::Write` cannot be known at compilation time
+ --> src/lib.rs:6:13
+  |
+6 |     generic(w);
+  |             ^ doesn't have a size known at compile-time
+...
+9 | fn generic<W: Write>(_w: &W) {}
+  |    ------- -       - help: consider relaxing the implicit `Sized` restriction: `+  ?Sized`
+  |            |
+  |            required by this bound in `generic`
+  |
+  = help: the trait `std::marker::Sized` is not implemented for `dyn std::io::Write`
+  = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0277`.
+error: could not compile `error-messages`.
+
+To learn more, run the command again with --verbose.
+```
+
+</details>
+
+![A terminal screenshot of the 1.43.0 error message.][trait-error-1.43.0]
+
+##### Second Example (help)
+```rust
+fn main() {
+    let s = "".to_owned();
+    println!("{:?}", s.find("".to_owned()));
+}
+```
+
+<details>
+ <summary>1.2.0 Error Message</summary>
+
+```
+   Compiling error-messages v0.1.0 (file:///Users/ep/src/rust/error-messages)
+src/lib.rs:3:24: 3:43 error: the trait `core::ops::FnMut<(char,)>` is not implemented for the type `collections::string::String` [E0277]
+src/lib.rs:3     println!("{:?}", s.find("".to_owned()));
+                                    ^~~~~~~~~~~~~~~~~~~
+note: in expansion of format_args!
+<std macros>:2:25: 2:56 note: expansion site
+<std macros>:1:1: 2:62 note: in expansion of print!
+<std macros>:3:1: 3:54 note: expansion site
+<std macros>:1:1: 3:58 note: in expansion of println!
+src/lib.rs:3:5: 3:45 note: expansion site
+src/lib.rs:3:24: 3:43 error: the trait `core::ops::FnOnce<(char,)>` is not implemented for the type `collections::string::String` [E0277]
+src/lib.rs:3     println!("{:?}", s.find("".to_owned()));
+                                    ^~~~~~~~~~~~~~~~~~~
+note: in expansion of format_args!
+<std macros>:2:25: 2:56 note: expansion site
+<std macros>:1:1: 2:62 note: in expansion of print!
+<std macros>:3:1: 3:54 note: expansion site
+<std macros>:1:1: 3:58 note: in expansion of println!
+src/lib.rs:3:5: 3:45 note: expansion site
+error: aborting due to 2 previous errors
+Could not compile `error-messages`.
+
+To learn more, run the command again with --verbose.
```

Comment:
Do we need text and images for the error messages? Seems a bit like overkill

Reply:
@nrc It's a compromise, as neither representation is ideal.

The text based versions don't show the work that has gone into colour and emphasis as part of the error message, and have the problem on mobile where you can only see half (vertically) of the text at a time.

<img width="386" alt="Screenshot 2020-05-06 at 09 24 49" src="https://user-images.githubusercontent.com/4464295/81147905-88f92480-8f7b-11ea-90d0-e8dd3fd0ae50.png">

The images do show the colour font work, but can also be hard to read on mobile because the images are landscape and most view the web in portrait. Also if you rely on a screenreader you would otherwise have no way of telling what the error message looks like. Having both lets the reader choose the representation that most readable best for them.

<img width="343" alt="Screenshot 2020-05-06 at 09 22 55" src="https://user-images.githubusercontent.com/4464295/81148378-76331f80-8f7c-11ea-9bb5-24c02cb14058.png">